samcli/lib/deploy/deployer.py

```diff
@@ -642,6 +675,104 @@ def get_stack_outputs(self, stack_name, echo=True):
         except botocore.exceptions.ClientError as ex:
             raise DeployStackOutPutFailedError(stack_name=stack_name, msg=str(ex)) from ex
 
+    def rollback_delete_stack(self, stack_name: str):
+        """
+        Try to rollback the stack to a sucessful state, if there is no good state then delete the stack
+
+        Parameters
+        ----------
+        :param stack_name: str
+            The name of the stack
+        """
+        kwargs = {
+            "StackName": stack_name,
+        }
+
+        current_state = self._get_stack_status(stack_name)
+
+        try:
+            if current_state == "UPDATE_FAILED":
+                LOG.info("Stack %s failed to update, rolling back stack to previous state...", stack_name)
+
+                self._client.rollback_stack(**kwargs)
+                self.describe_stack_events(stack_name, time.time() * 1000, FailureMode.DELETE)
+                self._rollback_wait(stack_name)
+
+                current_state = self._get_stack_status(stack_name)
+
+            failed_states = ["CREATE_FAILED", "UPDATE_FAILED", "ROLLBACK_COMPLETE", "ROLLBACK_FAILED"]
+
+            if current_state in failed_states:
+                LOG.info("Stack %s failed to create/update correctly, deleting stack", stack_name)
+
+                self._client.delete_stack(**kwargs)
```

Comment:
Add a case to check for DELETE_FAILED, where we were not able to delete some resources (eg S3 bucket) and let them know that they need to clean it up.